crates/nu-protocol/src/pipeline_data/list_stream.rs

```diff
@@ -0,0 +1,139 @@
+use crate::{Config, PipelineData, ShellError, Span, Value};
+use std::{
+    fmt::Debug,
+    sync::{atomic::AtomicBool, Arc},
+};
+
+pub type ValueIterator = Box<dyn Iterator<Item = Value> + Send + 'static>;
+
+/// A potentially infinite stream of values, optionally with a mean to send a Ctrl-C signal to stop
+/// the stream from continuing.
+///
+/// In practice, a "stream" here means anything which can be iterated and produce Values as it iterates.
+/// Like other iterators in Rust, observing values from this stream will drain the items as you view them
+/// and the stream cannot be replayed.
+pub struct ListStream {
+    stream: ValueIterator,
+    span: Span,
+}
```

Comment:
What is the motivation to move out the ctrlc handling from the old implementation

Reply:
The old implementation would often use the `ListStream` as an iterator and then turn it back into a `ListStream` via `into_pipline_data(ctrlc)`. This would add a ctrlc handler to every `dyn` level in the iterator. In this PR, I added `ListStream::map` and `ListStream::modify` which operate directly on `ListStream`s. If I kept the ctrlc in `ListStream`, this would check the ctrlc only after potentially iterating through multiple items in the inner iterator. So, I needed the ctrlc handler to be kept at the initial iterator / the source. I used the `Interrupt` struct for this which made it unnecessary to store the ctrlc in ListStream.